### PR TITLE
v4 FillQuoteTransformer

### PR DESCRIPTION
**Pull request by dorothy-zbornak**
## Description

Unabashed ripoff of #83 so I can take all the glory

- Supports V4 native order types (limit + RFQ).
- No V3 order type support :wave:.
- Uses a new, simpler order type/struct for bridge "orders".
  - We no longer support legacy bridge contracts. Now use LiquidityProvider contracts for non-rolled-up sources.
  - LiquidityProvider `bridgeData` is doubly-encoded: `abi.encode(lpAddress, abi.encode(...bridgeDataParams))`
  - `bridgeData` for some rolled-up sources have changed slightly. Basically removed anywhere we had to encode the `sellToken` because we know that.
- `ERC20BridgeTransfer` is now `BridgeFill`, which takes slightly different params and is fired for all bridge fills (including LP fills).
- Support for proportional `fillAmount` values (in sells) to sell a % of the taker token balance: `1<<255 | (FRACTION * 1e18)`

### TODO
- [x] tests
- [x] merge https://github.com/0xProject/tools/pull/16
- [ ] merge this into a working branch

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/utils/contracts/src/v06/errors/LibSafeMathRichErrorsV06.sol [hunk 1]
```diff
@@ -39,7 +39,8 @@ library LibSafeMathRichErrorsV06 {
     enum DowncastErrorCodes {
         VALUE_TOO_LARGE_TO_DOWNCAST_TO_UINT32,
         VALUE_TOO_LARGE_TO_DOWNCAST_TO_UINT64,
-        VALUE_TOO_LARGE_TO_DOWNCAST_TO_UINT96
+        VALUE_TOO_LARGE_TO_DOWNCAST_TO_UINT96,
+        VALUE_TOO_LARGE_TO_DOWNCAST_TO_UINT128
```
Comment (dorothy-zbornak) [inline contracts/utils/contracts/src/v06/errors/LibSafeMathRichErrorsV06.sol:43]: https://github.com/0xProject/tools/pull/16

File: contracts/zero-ex/test/transformers/fill_quote_transformer_test.ts [hunk 2]
```diff
@@ -5,16 +5,28 @@ import {
     expect,
     getRandomInteger,
     Numberish,
-    randomAddress,
 } from '@0x/contracts-test-utils';
-import { assetDataUtils } from '@0x/order-utils';
-import { encodeFillQuoteTransformerData, FillQuoteTransformerData, FillQuoteTransformerSide } from '@0x/protocol-utils';
-import { Order } from '@0x/types';
+import {
```
Comment (dorothy-zbornak) [inline contracts/zero-ex/test/transformers/fill_quote_transformer_test.ts:9]: Basically rewrote this whole file, so probably not worth looking at the diff.

File: contracts/zero-ex/contracts/src/transformers/bridges/BridgeSource.sol [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    uint256 constant internal LIQUIDITYPROVIDER = 8;
+    uint256 constant internal MOONISWAP = 9;
+    uint256 constant internal MSTABLE = 10;
+    uint256 constant internal OASIS = 11;
+    uint256 constant internal SHELL = 12;
+    uint256 constant internal SNOWSWAP = 13;
+    uint256 constant internal SUSHISWAP = 14;
+    uint256 constant internal SWERVE = 15;
+    uint256 constant internal UNISWAP = 16;
+    uint256 constant internal UNISWAPV2 = 17;
```
Comment (dorothy-zbornak) [inline contracts/zero-ex/contracts/src/transformers/bridges/BridgeSource.sol:42]: Just use the next available number for new, non-rolled up sources. Or any random unused number, who cares.

File: contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinBancor.sol [hunk 4]
```diff
@@ -64,17 +64,22 @@ contract MixinBancor is
...
+        IERC20TokenV06[] memory path;
+        IBancorNetwork bancorNetworkAddress;
+        {
+            address[] memory _path;
+            (
+                _path,
+                bancorNetworkAddress
+            ) = abi.decode(bridgeData, (address[], IBancorNetwork));
+            // To get around `abi.decode()` not supporting interface array types.
+            assembly { path := _path }
```
Comment (dorothy-zbornak) [inline contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinBancor.sol:76]: I bet this works on 0.8

File: contracts/zero-ex/test/transformers/fill_quote_transformer_test.ts [hunk 5]
```diff
@@ -265,875 +357,1013 @@ blockchainTests.resets('FillQuoteTransformer', env => {
...
-                    encodeTransformData({
-                        orders,
-                        signatures,
-                    }),
-                )
-                .awaitTransactionSuccessAsync({ value: ZERO_AMOUNT });
-            assertBalances(await getBalancesAsync(host.address), {
-                ...ZERO_BALANCES,
-                makerAssetBalance: qfr.makerAssetBought,
+        it('can fully sell to a single RFQ order', async () => {
```
Comment (mzhu25) [inline contracts/zero-ex/test/transformers/fill_quote_transformer_test.ts:1128]:
```suggestion
        it('can fully buy to a single RFQ order', async () => {
```

File: contracts/zero-ex/test/transformers/fill_quote_transformer_test.ts [hunk 6]
```diff
@@ -265,875 +357,1013 @@ blockchainTests.resets('FillQuoteTransformer', env => {
...
-                    encodeTransformData({
-                        orders,
-                        signatures,
-                    }),
-                )
-                .awaitTransactionSuccessAsync({ value: singleProtocolFee.times(nativeOrders.length) });
-            assertBalances(await getBalancesAsync(host.address), {
-                ...ZERO_BALANCES,
-                makerAssetBalance: qfr.makerAssetBought,
+        it('can partially sell to a single RFQ order', async () => {
```
Comment (mzhu25) [inline contracts/zero-ex/test/transformers/fill_quote_transformer_test.ts:1152]:
```suggestion
        it('can partially buy to a single RFQ order', async () => {
```

**Review by mzhu25 — APPROVED** [review]
Couldn't have written it better myself

File: contracts/zero-ex/contracts/src/transformers/bridges/IBridgeAdapter.sol [hunk 7]
```diff
@@ -18,12 +18,51 @@
...
 interface IBridgeAdapter {
 
+    struct BridgeOrder {
+        uint256 source;
+        uint256 takerTokenAmount;
+        uint256 makerTokenAmount;
+        bytes bridgeData;
+    }
+
+    struct Addresses {
```
Comment (dekz) [inline contracts/zero-ex/contracts/src/transformers/bridges/IBridgeAdapter.sol:35]: should we just force these into `bridgeData`? Apart from WETH

File: packages/protocol-utils/src/transformer_utils.ts [hunk 8]
```diff
@@ -52,19 +80,91 @@ export enum FillQuoteTransformerSide {
...
+    fillSequence: FillQuoteTransformerOrderType[];
     fillAmount: BigNumber;
     refundReceiver: string;
-    rfqtTakerAddress: string;
+}
+
+/**
+ * Identifies the DEX type of a bridge order.
+ */
+export enum BridgeSource {
```
Comment (dekz) [inline packages/protocol-utils/src/transformer_utils.ts:120]: Should comment that this should be append only, or we explicitly set the value. One might assume a new source should be added alphabetically

**Review by dekz — APPROVED** [review]

File: contracts/zero-ex/contracts/src/transformers/FillQuoteTransformer.sol [hunk 9]
```diff
@@ -276,223 +284,174 @@ contract FillQuoteTransformer is
...
+        (bool success, bytes memory resultData) = address(bridgeAdapter).delegatecall(
+            abi.encodeWithSelector(
+                IBridgeAdapter.trade.selector,
+                order,
+                data.sellToken,
+                data.buyToken,
+                takerTokenFillAmount
+            )
         );
+        if (success) {
```
[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/zero-ex/contracts/src/transformers/FillQuoteTransformer.sol:313]: idk, maybe we'll pad certain bridges some day. Is the concern about making reverts cheaper?

File: contracts/zero-ex/contracts/src/transformers/bridges/IBridgeAdapter.sol (hunk 7 above)
Comment (dorothy-zbornak) [reply contracts/zero-ex/contracts/src/transformers/bridges/IBridgeAdapter.sol:35]: I've thought about it. I wasn't sure if it was a good idea to change the bridge data encoding too much.

File: contracts/zero-ex/contracts/src/transformers/FillQuoteTransformer.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/zero-ex/contracts/src/transformers/FillQuoteTransformer.sol:313]: Guess it does make FQT overrides in simbot easier?

[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/zero-ex/contracts/src/transformers/FillQuoteTransformer.sol:313]: OK, you've got it.

File: contracts/zero-ex/contracts/src/transformers/bridges/IBridgeAdapter.sol (hunk 7 above)
Comment (dorothy-zbornak) [reply contracts/zero-ex/contracts/src/transformers/bridges/IBridgeAdapter.sol:35]: Guess it does make FQT overrides in simbot easier?

Comment (dorothy-zbornak) [reply contracts/zero-ex/contracts/src/transformers/bridges/IBridgeAdapter.sol:35]: I did it

**Pull request merged**